### ISSUE TITLE
Model return NaN predictions in larger datasets

### ISSUE THREAD

**Issue by dbeinhauer** [dbeinhauer/mcs-source#47]
In case we wanted to use larger model or smaller time steps the model starts to return NaN predictions. We need to find out what it the source of it and try to remove it to always get reasonable results.

This problem is also visible in case we clip the weights manually (inhibitory/excitatory layers). Usually, the problem is while the model predicts larger and larger values (in final evaluation it leads to infinity (longer sequence to infer)).

The possible problems might be exploding gradients, large learning rate, ReLU activation (not bounded) and weights clipping.

**Comment by dbeinhauer** [comment]
We will try to implement gradient clipping to resolve possible exploding gradients issue.

Other possibility is also using other activation function than ReLU. The thing is that we need non-negative predictions but they are quite strict. It predicts number of spikes in given time interval (20 ms). The thing is that there happens only very limited number of spikes (max 4 in the training data), so, the possible solution might be using some bounded ReLU alternative (sigmoid, tanh, softmax etc.).

**Comment by dbeinhauer** [comment]
After inspection of the gradients it definitely looks like the exploding gradients are the case. In non problematic variants gradients reached the values over 0.1 in very rare occasions. On the other hand during problematic learning it reached sometimes values over 600.

So, the first solution now would be application of the gradient clipping (probably to range 1). Additionally, using some bounded activation function would be also beneficial.

**Comment by dbeinhauer** [comment]
The problem was probably mainly caused by exploding gradients. We try to eliminate this using gradient clipping. We also do use bounded ReLU to limit the model to predict only up to 20 at the DNN of the neuron (because it it the highest possible response of the model that does make sense).
